tests/unit/s2n_tls13_new_session_ticket_test.c

```diff
@@ -0,0 +1,188 @@
+    uint8_t *server_implicit_iv = conn->server->server_implicit_iv;
+    uint8_t *client_implicit_iv = conn->client->client_implicit_iv;
+ 
+    /* Initialize record algorithm */
+    POSIX_GUARD(cipher_suite->record_alg->cipher->init(server_session_key));
+    POSIX_GUARD(cipher_suite->record_alg->cipher->init(client_session_key));
+    POSIX_GUARD(cipher_suite->record_alg->cipher->set_encryption_key(server_session_key, &key));
+    POSIX_GUARD(cipher_suite->record_alg->cipher->set_encryption_key(client_session_key, &key));
+    POSIX_GUARD(cipher_suite->record_alg->cipher->set_decryption_key(server_session_key, &key));
+    POSIX_GUARD(cipher_suite->record_alg->cipher->set_decryption_key(client_session_key, &key));
+
+    /* Initialized secrets */
+    POSIX_CHECKED_MEMCPY(conn->secure.server_app_secret, application_secret.data, application_secret.size);
+    POSIX_CHECKED_MEMCPY(conn->secure.client_app_secret, application_secret.data, application_secret.size);
+ 
+    /* Copy iv bytes from input data */
+    POSIX_CHECKED_MEMCPY(server_implicit_iv, iv.data, iv.size);
+    POSIX_CHECKED_MEMCPY(client_implicit_iv, iv.data, iv.size);
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_setup_test_ticket_key(struct s2n_config *config)
+{
+    POSIX_ENSURE_REF(config);
+
+    /**
+     *= https://tools.ietf.org/rfc/rfc5869#appendix-A.1
+     *# PRK  = 0x077709362c2e32df0ddc3f0dc47bba63
+     *#        90b6c73bb50f9c3122ec844ad7c2b3e5 (32 octets)
+     **/
+    S2N_BLOB_FROM_HEX(ticket_key,
+    "077709362c2e32df0ddc3f0dc47bba63"
+    "90b6c73bb50f9c3122ec844ad7c2b3e5");
+
+    /* Set up encryption key */
+    uint64_t current_time;
+    uint8_t ticket_key_name[16] = "2016.07.26.15\0";
+    EXPECT_SUCCESS(config->wall_clock(config->sys_clock_ctx, &current_time));
+    EXPECT_SUCCESS(s2n_config_add_ticket_crypto_key(config, ticket_key_name, strlen((char *)ticket_key_name),
+                    ticket_key.data, ticket_key.size, current_time/ONE_SEC_IN_NANOS));
+
+    return S2N_SUCCESS;
+}
+
+int main(int argc, char **argv)
+{   
+    BEGIN_TEST();
+    
+    /* s2n_send sends NewSessionTicket message */
+    {
+        struct s2n_connection *server_conn= s2n_connection_new(S2N_SERVER);
+        struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT);
+        EXPECT_NOT_NULL(server_conn);
+        EXPECT_NOT_NULL(client_conn);
+
+        struct s2n_config *config = s2n_config_new();
+        EXPECT_NOT_NULL(config);
+
+        EXPECT_SUCCESS(s2n_config_set_session_tickets_onoff(config, 1));
+        EXPECT_SUCCESS(s2n_setup_test_ticket_key(config));
+        EXPECT_SUCCESS(s2n_config_set_session_ticket_cb(config, s2n_test_session_ticket_cb, NULL));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+
+        EXPECT_SUCCESS(s2n_test_init_encryption(server_conn));
+        EXPECT_SUCCESS(s2n_test_init_encryption(client_conn));
+
```

Comment:
Why not perform the handshake properly before testing the NST, using s2n_negotiate? Then there won't be any chance that you're not setting things up the way a connection sets itself up.

Reply:
Good idea